fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/AvroDimensionRowParser.java

```diff
@@ -0,0 +1,185 @@
+
+        // Returns true, only if all the dimension fields present in the configured dimension exist in the AVRO schema
+        return convertedDimensionFields.values().containsAll(avroFields);
+    }
+
+    /**
+     * Parses the avro file and populates the dimension rows after validating the schema.
+     *
+     * @param dimension The dimension object used to configure the dimension
+     * @param avroFilePath The path of the AVRO data file (.avro)
+     * @param avroSchemaPath The path of the AVRO schema file (.avsc)
+     *
+     * @throws NullPointerException thrown if JSON object `fields` is not present
+     * @throws IOException thrown if there is error parsing the avro files
+     */
+    public void parseAvroFileDimensionRows(Dimension dimension, String avroFilePath, String avroSchemaPath)
+        throws NullPointerException, IOException {
+
+        if (isSchemaValid(dimension, avroSchemaPath)) {
+
+            JsonNode jsonAvroSchema = convertFileToJsonNode(avroSchemaPath);
+
+            // Creates an AVRO schema object based on the parsed AVRO schema file (.avsc)
+            Schema schema = new Schema.Parser().parse(jsonAvroSchema.toString());
+
+            // Creates an AVRO DatumReader object based on the AVRO schema object
+            DatumReader<GenericRecord> datumReader = new GenericDatumReader<>(schema);
+
+            // Creates an AVRO DataFileReader object that reads the AVRO data file one record at a time
+            DataFileReader<GenericRecord> dataFileReader =
+                new DataFileReader<>(new File(avroFilePath), datumReader);
+
+            // Represents each record in the avro data file
+            GenericRecord avroData = null;
+
+            while (dataFileReader.hasNext()) {
+                avroData = dataFileReader.next(avroData);
+
+                // Creates a JsonObject from the AVRO GenericRecord object
+                JsonNode avroDataJson = MAPPER.readTree(new JsonFactory().createParser(avroData.toString()));
+
+                // Constructs a Map of the converted dimension field name and the corresponding value in the avro file
+                Map<String, String> dimensionFieldKeyValues = dimension.getDimensionFields()
+                        .stream()
+                        .collect(Collectors.toMap(dimensionField -> dimensionField.getName(),
+                                                  // Looks up converted field name from the `convertedDimensionFields`
+                                                  // Parsed as text to avoid the extra quotes around string objects
+                                                  dimensionField -> avroDataJson
+                                                        .get(convertedDimensionFields
+                                                                .get(dimensionField.getName())
+                                                        ).asText()
+                                                )
```

Comment:
This feels like we're doing way too many repeated steps in the tight loop.  Perhaps the necessary mapping needed to pull the details off the avro record can be done in advance so that it's a sequence of point lookups only inside the big loop.